lib/vrl/stdlib/src/map_keys.rs

```diff
@@ -0,0 +1,116 @@
+use vrl::prelude::*;
+
+#[derive(Clone, Copy, Debug)]
+pub struct MapKeys;
+
+impl Function for MapKeys {
+    fn identifier(&self) -> &'static str {
+        "map_keys"
+    }
+
+    fn parameters(&self) -> &'static [Parameter] {
+        &[
+            Parameter {
+                keyword: "value",
+                kind: kind::OBJECT,
+                required: true,
+            },
+            Parameter {
+                keyword: "recursive",
+                kind: kind::BOOLEAN,
+                required: false,
+            },
+        ]
+    }
+
+    fn examples(&self) -> &'static [Example] {
+        &[
+            Example {
+                title: "map object keys",
+                source: r#"map_keys({ "a": 1, "b": 2 }) -> |key| { upcase(key) }"#,
+                result: Ok(r#"{ "A": 1, "B": 2 }"#),
+            },
+            Example {
+                title: "recursively map object keys",
+                source: r#"map_keys({ "a": 1, "b": [{ "c": 2 }, { "d": 3 }], "e": { "f": 4 } }, recursive: true) -> |key| { upcase(key) }"#,
+                result: Ok(r#"{ "A": 1, "B": [{ "C": 2 }, { "D": 3 }], "E": { "F": 4 } }"#),
+            },
+        ]
+    }
+
+    fn compile(
+        &self,
+        _state: (&mut state::LocalEnv, &mut state::ExternalEnv),
+        _ctx: &mut FunctionCompileContext,
+        mut arguments: ArgumentList,
+    ) -> Compiled {
+        let value = arguments.required("value");
+        let recursive = arguments.optional("recursive");
+        let closure = arguments.required_closure()?;
+
+        Ok(Box::new(MapKeysFn {
+            value,
+            closure,
+            recursive,
+        }))
+    }
+
+    fn closure(&self) -> Option<closure::Definition> {
+        use closure::{Definition, Input, Output, Variable, VariableKind};
+
+        Some(Definition {
+            inputs: vec![Input {
+                parameter_keyword: "value",
+                kind: Kind::object(Collection::any()),
+                variables: vec![Variable {
+                    kind: VariableKind::Exact(Kind::bytes()),
+                }],
+                output: Output::Kind(Kind::bytes()),
+                example: Example {
```

Comment:
This could effectively just be `self.examples[0]` to avoid a bit of duplication?

Reply:
It could, but this example is actually used in diagnostic messages, which the `self.examples` is not. Additionally, this example is tied to the input variant. Granted, there's only one right now, but that might change.

I don't mind removing the duplication, but it has a tiny risk of causing confusion if we ever change the `self.examples` example, such that it no longer matches this input definition, and people are presented with that erroneous example in diagnostic messages.

Reply:
Had the same question in the other PRs. Didn't know either that it's tied to diagnostic messages, makes sense.

Are the examples for diagnostic messages itself run as tests as well? 😄

Reply:
They are, test _all_ the things 😜 

I added that here: https://github.com/vectordotdev/vector/pull/12338/files#diff-38054ba10e63cb6b2e79e1393ac1075267b652e063d34354f259417228093fed